flash-patcher.py

```diff
@@ -220,293 +351,450 @@ def apply_patch(patch_file):
-                    perror("Line number " + str(line_end) + " out of range for file " + file_location)
-                    perror("Aborting...")
-                    exit(1)
-
-                write_to_file(file_location, current_file)
+            # Account for spaces in file name.
+            # Take everything except the first and last blocks.
+            # The first block is the command character.
+            # The last block is the line number(s).
+            short_name = " ".join(split_line[1:-1])
+            file_location = DECOMP_LOCATION_WITH_SCRIPTS / short_name
+
+            # Add the current script to the list of modified ones,
+            # i.e. keep this in the final output.
+            modified_scripts.add(file_location)
+
+            current_file = read_from_file(file_location, patch_file, current_line_no)
+            line_counts = split_line[-1].split("-")
+
+            if len(line_counts) != 2:
+                exception(
+                    """%s, line %d: Invalid syntax.
+                    Expected two integers, separated by a dash (-) (at %s)""",
+                    patch_file,
+                    current_line_no,
+                    line_stripped,
+                )
+                sys.exit(1)
+
+            try:
+                line_start = int(line_counts[0])
+                line_end = int(line_counts[1])
+            except ValueError:
+                exception(
+                    """%s, line %d: Invalid syntax.
+                    Invalid line numbers provided: %s
+                    Aborting...""",
+                    patch_file,
+                    current_line_no,
+                    split_line[-1],
+                )
+                sys.exit(1)
+
+            try:
+                for _i in range(line_start, line_end + 1):
+                    del current_file[line_start - 1]
+            except IndexError:
+                exception(
+                    """%s, line %d: Out of range.
+                    Line number %d out of range for file %s.
+                    Aborting...""",
+                    patch_file,
+                    current_line_no,
+                    line_end,
+                    file_location,
+                )
+                sys.exit(1)
+
+            write_to_file(file_location, current_file)
 
         # Unrecognized statement
         else:
-            perror("Unrecognized command: '" + split_line[0] + "', skipping (at " + patch_file + ", line " + str(current_line_no) + ")")
+            error(
+                "Unrecognized command: '%s', skipping (at %s, line %d)",
+                split_line[0],
+                patch_file,
+                current_line_no,
+            )
 
         current_line_no += 1
 
     if line_add_mode:
-        perror("")
-        perror(patch_file + ": Syntax error")
-        perror("Missing end-patch for \"add\" on line " + str(injector.startingLineNo - 1))
-        perror("Aborting...")
-        exit(1)
+        critical(
```

Comment:
When should we use error vs. critical? If both cause a program abort but no system/external corruption, error is probably the better choice

Reply:
I've been using `critical` to indicate an abort, `error` to indicate an error without abort (i.e, invalid input but not so serious that we have to abort), and theoretically `warn` to mean something that should be changed but isn't inherently a problem.

Reply:
We should use exceptions whenever the program reaches an error that requires immediate exit. I feel like critical should only be used if we cause permanent issues to the system (for example, a force-reboot that may lead to permanent file corruption).

We should also avoid using error if the program doesn't require an immediate exit (use warn instead) since in most contexts, an error would require the program to stop immediately. 
(Alternatively - and admittedly this is a long term solution - we could use a syntax checker *before* the main injection pass, that could check for this type of error and exit immediately)

Reply:
Exception is meant to be used in `except` blocks only, see https://docs.python.org/3/library/logging.html#logging.exception

Reply:
ah, understood. We should still use `warning` instead of error for any message that doesn't cause the program to exit imo.
(Also, this message should be switched to error)